org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/BatchImageBuilder.java

```diff
@@ -348,4 +363,50 @@ protected void storeProblemsFor(SourceFile sourceFile, CategorizedProblem[] prob
 public String toString() {
 	return "batch image builder for:\n\tnew state: " + this.newState; //$NON-NLS-1$
 }
+
+@Override
+public void startBatch() {
+	this.batchMode = true;
+}
+
+@Override
+protected void writeClassFileContents(ClassFile classFile, IFile file, String qualifiedFileName, boolean isTopLevelType,
+		SourceFile compilationUnit) throws CoreException {
+	if (JavaBuilder.DEBUG) {
```

Comment:
This is not ambiguous and should be split into two traces: **writing** file and **caching** file contents for later write